core/modules/memman/MemFile.cc

```diff
@@ -0,0 +1,181 @@
+
+    MemInfo mInfo;
+    std::lock_guard<std::mutex> gaurd(cacheMutex);
+
+    // If the file is already locked, indicate success
+    //
+    if (_isLocked) {
+        if (_isFlex) _memory.flexNum(1);
+        MLResult aokResult(_memInfo.size(), 0);
+        return aokResult;
+    }
+
+    // If the file doesn't meet the refcount restrictio, don't lock it
+    //
+    if (_refs < minRefs) {
+        MLResult nilResult(0,0);
+        return nilResult;
+    }
+
+    // If a space is wanted, do so now before we attemptto lock the file
+    //
+    if (maxBytes != 0 && _memInfo.size() > maxBytes) {
+        MLResult bigResult(0, ENOMEM);
+        return bigResult;
+    }
+
+    // Lock this table in memory if possible.
+    //
+    mInfo = _memory.memLock(_fPath, _isFlex);
+
+    // If we successfully locked this file, then indicate so, update the
+    // memory information and return.
+    //
+    if (mInfo.isValid()) {
+        MLResult aokResult(mInfo.size(),0);
+        _isLocked = 1;
+        _memInfo = mInfo;
+        return aokResult;
+    }
+
+    // Diagnose any errors
+    //
+    MLResult errResult(0, mInfo.errCode());
+    return errResult;
+}
+
+/******************************************************************************/
+/*                              n u m F i l e s                               */
+/******************************************************************************/
+
+uint32_t MemFile::numFiles() {
+
+    std::lock_guard<std::mutex> gaurd(cacheMutex);
+
+    // Simply return the size of our file cache
+    //
+    return fileCache.size();
+}
+
+/******************************************************************************/
+/*                                o b t a i n                                 */
+/******************************************************************************/
+  
+MemFile::MFResult MemFile::obtain(std::string const& fPath,
+                                  Memory& mem, bool isFlex) {
+
+    MemFile *mfP;
+    MemInfo  mInfo;
+    std::lock_guard<std::mutex> gaurd(cacheMutex);
+
+    // First look up if this table already exists in our cache and is using the
+    // the same memory object (error if not). If so, up the reference count and
+    // return the object as it may be shared. Note: it->second == MemFile*!
+    //
+    auto it = fileCache.find(fPath);
+    if (it != fileCache.end()) {
+        if (&(it->second->_memory) != &mem) {
+            MFResult errResult(0, EXDEV);
+            return errResult;
+        }
+        it->second->_refs++;
+        MFResult aokResult(&(*(it->second)),0);
+        return aokResult;
+    }
+
+    // Validate the file and get its size
+    //
+    mInfo = mem.fileInfo(fPath);
+    if (!mInfo.isValid()) {
+        MFResult errResult(0, mInfo.errCode());
+        return errResult;
+    }
+
+    // Get a new file object and insert it into the map
+    //
+    mfP = new MemFile(fPath, mem, mInfo, isFlex);
+    fileCache.insert({fPath, mfP});
```

Comment:
Right, `insert` is overloaded method with multiple signatures. How does compiler know that construct `{fPath, mfP}` is an `std::pair` and not anything else? Are there any possible surprises if you omit the actual type of that initializer? I guess I'm just to used to always providing explicit `value_type()`. For cases like this I'd probably prefer also-modern `fileCache.emplace(fPath, mfP);`